### PR TITLE
IDScopeProvider called before reaching out to Azure DPS.

### PR DESCRIPTION
**Pull request by stoyan-zoubev**
[#14] IDScopeProvider called too late

Signed-off-by: Stoyan Zoubev <Stoyan.Zoubev@bosch.io>

### CODE REVIEW

**Review by hristobojilov — APPROVED** [review]

File: config/connection_settings.go [hunk 1]
```diff
@@ -133,14 +139,27 @@ func CreateAzureCertificateConnectionSettings(
 	return connSettings, nil
 }
 
-// CreateAzureProvisioningConnectionSettings creates the configuration data for establishing connection to Azure device that requires device provisioning.
+// CreateAzureProvisioningConnectionSettings creates the configuration data for establishing connection to Azure device that requires device provisioning, using no IDScopeProvider.
```
Comment (e-grigorov) [inline config/connection_settings.go:142]: typo `connection to Azure device` should be `connection to Azure IoT Hub`

File: config/connection_settings.go [hunk 2]
```diff
@@ -133,14 +139,27 @@ func CreateAzureCertificateConnectionSettings(
...
 	provisioningService ProvisioningService,
 	provisioningFile io.ReadWriter,
 	useProvisioningClient bool,
 	certFileReader io.Reader,
 	keyFileReader io.Reader,
+) (*AzureConnectionSettings, error) {
+	return PrepareAzureProvisioningConnectionSettings(settings, nil, provisioningService, provisioningFile, useProvisioningClient, certFileReader, keyFileReader)
+}
+
+// PrepareAzureProvisioningConnectionSettings creates the configuration data for establishing connection to Azure device that requires device provisioning, allowing usage of IDScopeProvider.
```
Comment (e-grigorov) [inline config/connection_settings.go:154]: typo `connection to Azure device` should be `connection to Azure IoT Hub`

File: config/connection_settings_test.go [hunk 3]
```diff
@@ -451,9 +489,28 @@ func mockProvisioningService(t *testing.T, controller *gomock.Controller, device
 	return provisioningService
 }
 
+func mockProvisioningServiceWithIDScope(t *testing.T, controller *gomock.Controller, deviceData *config.AzureDeviceData, deviceDataError error, hasProvisioningFile bool, timesGetDataCalled, timesInitCalled int) *mock.MockProvisioningService {
+	provisioningService := mock.NewMockProvisioningService(controller)
+	provisioningService.EXPECT().GetDeviceData("dummyIdScope", gomock.Any()).Return(deviceData, deviceDataError).Times(timesGetDataCalled)
+	provisioningService.EXPECT().Init(gomock.Any(), gomock.Any()).Do(func(client config.ProvisioningHTTPClient, provisioningFile io.ReadWriter) {
+		if hasProvisioningFile != (client == nil) {
+			t.Fail()
```
Comment (e-grigorov) [inline config/connection_settings_test.go:497]: Here, this fail can happen in different combination i.e. with or without provisioning file. I would suggest to add a fail reason message.

File: config/connection_settings.go [hunk 4]
```diff
@@ -56,8 +56,13 @@ type AzureConnectionSettings struct {
 	*SharedAccessKey
 }
 
-// CreateAzureConnectionSettings creates the configuration data for establishing connection to the Azure IoT Hub.
+// CreateAzureConnectionSettings creates the configuration data for establishing connection to the Azure IoT Hub, using no IDScopeProvider.
```
Comment (e-grigorov) [inline config/connection_settings.go:59]: How do you use this method? Do we need it at all?

File: config/connection_settings.go [hunk 5]
```diff
@@ -56,8 +56,13 @@ type AzureConnectionSettings struct {
...
 func CreateAzureConnectionSettings(settings *AzureSettings, log logger.Logger) (*AzureConnectionSettings, error) {
+	return PrepareAzureConnectionSettings(settings, nil, log)
+}
+
+// PrepareAzureConnectionSettings creates the configuration data for establishing connection to the Azure IoT Hub, allowing usage of IDScopeProvider.
+func PrepareAzureConnectionSettings(settings *AzureSettings, idScopeProvider IDScopeProvider, log logger.Logger) (*AzureConnectionSettings, error) {
```
Comment (e-grigorov) [inline config/connection_settings.go:65]: If we keep Create, the name should be revised.

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: config/connection_settings.go (hunk 4 above)
Comment (stoyan-zoubev) [reply config/connection_settings.go:59]: This is the original method. I did not touch its signature as I did not want to go and modify all the tests too, as I wanted the PR to have a little a possible modified lines.
If you are OK with several extra lines with changes in the tests, then this method is not needed.

File: config/connection_settings.go (hunk 5 above)
Comment (stoyan-zoubev) [reply config/connection_settings.go:65]: See my comment from above, Create is the original method and it is not really needed.

File: config/connection_settings.go (hunk 1 above)
Comment (stoyan-zoubev) [reply config/connection_settings.go:142]: This typo is from the initial contribution of the source code, I suppose there are other typos too and I would suggest to open another ticket for fixing such issues.

File: config/connection_settings.go (hunk 2 above)
Comment (stoyan-zoubev) [reply config/connection_settings.go:154]: This typo is from the initial contribution source code, I suppose there are other typos too and I would suggest to open another ticket for fixing such issues.

File: config/connection_settings_test.go (hunk 3 above)
Comment (stoyan-zoubev) [reply config/connection_settings_test.go:497]: This method is a copy of another one from the initial contribution source code, I suppose there are other places that would need fail reason message and I would suggest to open another ticket for fixing such issues, not on just one place.

File: config/connection_settings.go (hunk 1 above)
Comment (e-grigorov) [reply config/connection_settings.go:142]: @stoyan-zoubev - any news here? Did you have time to check for some more typos? I cannot see new related issues in the project. In any case, I would suggest to fix what's found in the current PR.

File: config/connection_settings.go (hunk 2 above)
Comment (e-grigorov) [reply config/connection_settings.go:154]: @stoyan-zoubev - any news here? Did you have time to check for some more typos? I cannot see new related issues in the project. In any case, I would suggest to fix what's found in the current PR.

File: config/connection_settings_test.go (hunk 3 above)
Comment (e-grigorov) [reply config/connection_settings_test.go:497]: @stoyan-zoubev - any news here? Did you have time to check for similar issues? I cannot see new related issues in the project. In any case, I would suggest to fix what's found in the current PR.

File: config/connection_settings.go (hunk 4 above)
Comment (e-grigorov) [reply config/connection_settings.go:59]: I am not sure that I can understand your concerns here. In the end, we have to improve the code base, the number of modified lines should not be an issue.

**Comment by stoyan-zoubev** [comment]
Closing this PR as it is superseded by a new one - https://github.com/eclipse-kanto/azure-connector/pull/22

**Pull request closed without merging**